packages/governance/src/governor/extensions/governor_timelock_execution.cairo

```diff
@@ -0,0 +1,306 @@
+        +SRC5Component::HasComponent<TContractState>,
+        impl GovernorCoreExecution: HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of GovernorComponent::GovernorExecutionTrait<TContractState> {
+        /// See `GovernorComponent::GovernorExecutionTrait::state`.
+        fn state(
+            self: @GovernorComponentState<TContractState>, proposal_id: felt252
+        ) -> ProposalState {
+            let current_state = self._state(proposal_id);
+
+            if current_state != ProposalState::Queued {
+                return current_state;
+            }
+
+            let contract = self.get_contract();
+            let this_component = GovernorCoreExecution::get_component(contract);
+
+            let queue_id = this_component.Governor_timelock_ids.read(proposal_id);
+            let timelock_dispatcher = this_component.get_timelock_dispatcher();
+            let operation_state = timelock_dispatcher.get_operation_state(queue_id);
+
+            let is_timelock_operation_pending = operation_state == OperationState::Waiting
+                || operation_state == OperationState::Ready;
+            let is_timelock_operation_done = operation_state == OperationState::Done;
+
+            if is_timelock_operation_pending {
+                ProposalState::Queued
+            } else if is_timelock_operation_done {
+                // This can happen if the proposal is executed directly on the timelock.
+                ProposalState::Executed
+            } else {
+                // This can happen if the proposal is canceled directly on the timelock.
+                ProposalState::Canceled
+            }
+        }
+
+        /// See `GovernorComponent::GovernorExecutionTrait::executor`.
+        ///
+        /// In this module, the executor is the timelock controller.
+        fn executor(self: @GovernorComponentState<TContractState>) -> ContractAddress {
+            let contract = self.get_contract();
+            let this_component = GovernorCoreExecution::get_component(contract);
+
+            this_component.timelock()
+        }
+
+        /// See `GovernorComponent::GovernorExecutionTrait::execute_operations`.
+        ///
+        /// Runs the already queued proposal through the timelock.
+        fn execute_operations(
+            ref self: GovernorComponentState<TContractState>,
+            proposal_id: felt252,
+            calls: Span<Call>,
+            description_hash: felt252
+        ) {
+            let mut contract = self.get_contract_mut();
+            let mut this_component = GovernorCoreExecution::get_component_mut(ref contract);
+
+            let timelock_dispatcher = this_component.get_timelock_dispatcher();
+
+            timelock_dispatcher
+                .execute_batch(calls, 0, this_component.timelock_salt(description_hash));
+
+            // Cleanup
+            this_component.Governor_timelock_ids.write(proposal_id, 0);
```

Comment:
It may be a bit more secure to update the state prior to making an external call, guards against reentrancy

Reply:
Note that there are no checks associated with the value for this particular case. Even if we don't update the value it should be ok, since the state of the timelock proposal should be correct one even if we don't remove the id. Also, the target of the execute may want to verify the state of the proposal at some point, and the id being accessible can help with that.